File: solutions/elixir/say/1/lib/say.ex
defmodule Say do
  @doc """
  Translate a positive integer into English.
  """
  @zero "zero"
  @small_numbers %{
    1 => "one",
    2 => "two",
    3 => "three",
    4 => "four",
    5 => "five",
    6 => "six",
    7 => "seven",
    8 => "eight",
    9 => "nine",
    10 => "ten",
    11 => "eleven",
    12 => "twelve",
    13 => "thirteen",
    14 => "fourteen",
    15 => "fifteen",
    16 => "sixteen",
    17 => "seventeen",
    18 => "eighteen",
    19 => "nineteen"
  }

  @multiples_of_ten %{
    20 => "twenty",
    30 => "thirty",
    40 => "forty",
    50 => "fifty",
    60 => "sixty",
    70 => "seventy",
    80 => "eighty",
    90 => "ninety"
  }

  @short_scale %{
    3 => "thousand",
    6 => "million",
    9 => "billion"
  }

  @spec in_english(integer) :: {atom, String.t()}
  def in_english(number)
      when number < 0 or
             number > 999_999_999_999,
      do: {:error, "number is out of range"}

  def in_english(0),
    do: {:ok, @zero}

  def in_english(number),
    do: {:ok, in_english!(number)}

  defp in_english!(number)
       when number < 20,
       do: @small_numbers[number]

  defp in_english!(number)
       when number < 100 do
    number_of_tens = div(number, 10)
    multiple_of_ten = number_of_tens * 10

    in_english!(
      _prefix = @multiples_of_ten[multiple_of_ten],
      _remainder = number - multiple_of_ten,
      _separator = "-"
    )
  end

  defp in_english!(number)
       when number < 1000 do
    number_of_hundreds = div(number, 100)
    multiple_of_hundred = number_of_hundreds * 100

    in_english!(
      _prefix = "#{in_english!(number_of_hundreds)} hundred",
      _remainder = number - multiple_of_hundred
    )
  end

  defp in_english!(number) do
    log10 = trunc(:math.log10(number) / 3) * 3
    scale_to_10_power = round(:math.pow(10, log10))
    quotient = div(number, scale_to_10_power)

    in_english!(
      _prefix = "#{in_english!(quotient)} #{@short_scale[log10]}",
      _remainder = number - quotient * scale_to_10_power
    )
  end

  defp in_english!(prefix, remainder, separator \\ " ") do
    if remainder === 0,
      do: prefix,
      else: "#{prefix}#{separator}#{in_english!(remainder)}"
  end
end

File: solutions/java/say/1/src/main/java/Say.java
import java.util.Map;
import java.util.Objects;
import java.util.stream.IntStream;

public class Say {
	private static final String[] BELOW_TWENTY = {
			"zero", "one", "two", "three", "four", "five",
			"six", "seven", "eight", "nine", "ten",
			"eleven", "twelve", "thirteen", "fourteen", "fifteen",
			"sixteen", "seventeen", "eighteen", "nineteen"
	};

	private static final Map<Integer, String> TENS = Map.ofEntries(
			Map.entry(20, "twenty"),
			Map.entry(30, "thirty"),
			Map.entry(40, "forty"),
			Map.entry(50, "fifty"),
			Map.entry(60, "sixty"),
			Map.entry(70, "seventy"),
			Map.entry(80, "eighty"),
			Map.entry(90, "ninety")
	);

	private static final String[] PERIOD_NAMES = {"", "thousand", "million", "billion"};

	public String say(long number) {
		if (number < 0 || number > 999_999_999_999L) {
			throw new IllegalArgumentException("Number out of range");
		}

		return this.streamOverThousandsChunks(number);
	}

	private String streamOverThousandsChunks(long number) {
		return IntStream.range(0, PERIOD_NAMES.length)
				       .mapToObj(i -> chunkToWords(number, i))
				       .filter(Objects::nonNull)
				       .reduce(this::toProperWordOrder)
				       .orElse("zero");
	}

	private String chunkToWords(long number, int periodIndex) {
		double periodBase = Math.pow(1000, periodIndex);
		double periodValue = number / periodBase;
		int chunk = (int) (periodValue % 1000);
		return chunk == 0
				       ? null
				       : this.toWords(chunk, periodIndex);

	}

	private String toWords(int chunk, int periodIndex) {
		String chunkWords = this.convertChunk(chunk);
		String periodLabel = PERIOD_NAMES[periodIndex].isEmpty()
				                     ? ""
				                     : " " + PERIOD_NAMES[periodIndex];
		return chunkWords + periodLabel;
	}

	private String toProperWordOrder(String a, String b) {
		return b + " " + a;
	}

	private String convertChunk(int number) {
		StringBuilder stringBuilder = new StringBuilder();

		if (number >= 100) {
			number = this.appendHundreds(stringBuilder, number);
		}

		if (number >= 20) {
			this.appendTensAndUnits(stringBuilder, number);
		}
		else if (number > 0) {
			stringBuilder.append(BELOW_TWENTY[number]);
		}

		return stringBuilder.toString();
	}

	private int appendHundreds(StringBuilder stringBuilder, int number) {
		stringBuilder.append(BELOW_TWENTY[number / 100])
				.append(" hundred");
		number %= 100;
		if (number != 0) {
			stringBuilder.append(" ");
		}

		return number;
	}

	private void appendTensAndUnits(StringBuilder stringBuilder, int number) {
		int tensPart = (number / 10) * 10; // e.g., 42 → 40
		stringBuilder.append(TENS.get(tensPart));
		number %= 10;
		if (number != 0) {
			stringBuilder.append("-")
					.append(BELOW_TWENTY[number]);
		}
	}
}
